File: part-02/src/test/java/eu/luminis/workshop/smallsteps/logic/appservice/LegoStockQueriesTest.java
package eu.luminis.workshop.smallsteps.logic.appservice;

import eu.luminis.workshop.smallsteps.helpers.TestLegoStockRepository;
import eu.luminis.workshop.smallsteps.logic.domainmodel.valueobjects.LegoSetNumber;
import eu.luminis.workshop.smallsteps.logic.domainmodel.valueobjects.LegoStoreId;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.IncompleteReturn;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoBox;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.StockState;
import org.junit.jupiter.api.Test;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

public class LegoStockQueriesTest {
    private final LegoSetNumber millenniumFalcon = new LegoSetNumber(75192);
    private final LegoSetNumber atAt = new LegoSetNumber(75288);
    private final LegoSetNumber r2d2 = new LegoSetNumber(75308);
    private final LegoStoreId bussum = new LegoStoreId();

    @Test
    public void should_return_report_with_of_all_currently_missing_parts() {
        TestLegoStockRepository repository = new TestLegoStockRepository();
        repository.setInitialState(new StockState(
                null,
                null,
                null,
                List.of(
                        new LegoBox(millenniumFalcon, 42, Map.of("3022", 7, "20105", 1)),
                        new LegoBox(millenniumFalcon, 37, Map.of("3022", 3, "60581", 1)),
                        new LegoBox(atAt, 5, Map.of("3022", 2, "18674", 1)),
                        new LegoBox(r2d2, 8, Map.of("3666", 2, "64799", 1))
                ),
                null)
        );
        LegoStockQueries queries = new LegoStockQueries(
                repository);

        Map<String, Integer> pairs = queries.currentlyMissingPartsReport(bussum);

        assertThat(pairs).containsExactlyEntriesOf(new LinkedHashMap<>() {{
            put("3022", 12);
            put("3666", 2);
            put("18674", 1);
            put("20105", 1);
            put("60581", 1);
            put("64799", 1);
        }});
    }

    @Test
    public void should_return_report_with_of_all_historically_missing_parts() {
        TestLegoStockRepository repository = new TestLegoStockRepository();
        repository.setInitialState(new StockState(
                null,
                null,
                null,
                null,
                List.of(
                        new IncompleteReturn(millenniumFalcon, Map.of("3022", 7, "20105", 1)),
                        new IncompleteReturn(millenniumFalcon, Map.of("3022", 3, "60581", 1)),
                        new IncompleteReturn(atAt, Map.of("3022", 2, "18674", 1)),
                        new IncompleteReturn(r2d2, Map.of("3666", 2, "64799", 1))
                ))
        );
        LegoStockQueries queries = new LegoStockQueries(repository);

        Map<String, Integer> historicallyMostLostParts = queries.historicallyMostLostParts(bussum);

        assertThat(historicallyMostLostParts)
                .containsExactlyEntriesOf(new LinkedHashMap<>() {{
                    put("3022", 12);
                    put("3666", 2);
                    put("18674", 1);
                    put("20105", 1);
                    put("60581", 1);
                    put("64799", 1);
                }});
    }
}

File: part-02/src/test/kotlin/eu/luminis/workshop/smallsteps/logic/appService/LegoStockQueriesTest.kt
package eu.luminis.workshop.smallsteps.logic.appService

import eu.luminis.workshop.smallsteps.api.helper.TestLegoStockRepository
import eu.luminis.workshop.smallsteps.logic.domainModel.valueObjects.LegoSetNumber
import eu.luminis.workshop.smallsteps.logic.domainService.helper.TestUsers
import eu.luminis.workshop.smallsteps.logic.domainService.state.buildIncompleteReturn
import eu.luminis.workshop.smallsteps.logic.domainService.state.buildLegoBox
import eu.luminis.workshop.smallsteps.logic.domainService.state.StockState
import kotlin.test.Test
import kotlin.test.assertEquals

internal class LegoStockQueriesTest {
    private val millenniumFalcon = LegoSetNumber(75192)
    private val atAt = LegoSetNumber(75288)
    private val r2d2 = LegoSetNumber(75308)

    @Test
    fun `should return report with of all currently missing parts`() {
        // given
        val queries = LegoStockQueries(
            TestLegoStockRepository(
                StockState(
                    incompleteStock = listOf(
                        buildLegoBox(millenniumFalcon, 42, mapOf("3022" to 7, "20105" to 1)),
                        buildLegoBox(millenniumFalcon, 37, mapOf("3022" to 3, "60581" to 1)),
                        buildLegoBox(atAt, 5, mapOf("3022" to 2, "18674" to 1)),
                        buildLegoBox(r2d2, 8, mapOf("3666" to 2, "64799" to 1)),
                    )
                )
            )
        )

        // when
        val result = queries.currentlyMissingPartsReport(TestUsers.bussum)

        // then
        assertEquals(
            listOf(
                "3022" to 12,
                "3666" to 2,
                "18674" to 1,
                "20105" to 1,
                "60581" to 1,
                "64799" to 1
            ),
            result
        )
    }

    @Test
    fun `should return report of historically most lost parts`() {
        // given
        val queries = LegoStockQueries(
            TestLegoStockRepository(
                StockState(
                    incompleteReturnHistory = listOf(
                        buildIncompleteReturn(millenniumFalcon, mapOf("3022" to 7, "20105" to 1)),
                        buildIncompleteReturn(millenniumFalcon, mapOf("3022" to 3, "60581" to 1)),
                        buildIncompleteReturn(atAt, mapOf("3022" to 2, "18674" to 1)),
                        buildIncompleteReturn(r2d2, mapOf("3666" to 2, "64799" to 1)),
                    )
                )
            )
        )

        // when
        val result = queries.historicallyMostLostParts(TestUsers.bussum)

        // then
        assertEquals(
            listOf(
                "3022" to 12,
                "3666" to 2,
                "18674" to 1,
                "20105" to 1,
                "60581" to 1,
                "64799" to 1
            ),
            result
        )
    }
}